Add SYSTEM_ALERT_WINDOW and WRITE_SETTINGS support

shiraji opened the pull request:
see #50 

## TODO
- [x] Add SYSTEM_ALERT_WINDOW support 
- [x] Add WRITE_SETTINGS support
- [x] Refactor `addWithCheckBody`
- [x] Refactor `addResultCaseBody`
- [x] Add validation that does not mix annotations for `@NeedsPermission`
- [x] Add onActivityResult
- [x] Add test cases

shiraji commented:
Here we go, @hotchemi

There are some points you should review very carefully.

* `PERMISSION_UTILS`, `addWithCheckBody` and `addResultCaseBody` are moved to `BaseProcessorUnit`
* To make it consist, local variable `activity` is now gone
* Create mock `android.provider.Setting` for testing purpose.

hotchemi commented:
Sorry for the delay. I'll review.

hotchemi commented:
Ah sorry for the late. Your change is so huge so I'm gonna dive into the detail this night.

hotchemi commented on processor/src/test/java/permissions/dispatcher/processor/data/Source.java:
```diff
@@ -3594,11 +3561,10 @@ protected String getName() {
                     "           target.showCamera();",
                     "       } else {",
-                    "           if (PermissionUtils.shouldShowRequestPermissionRationale(activity, PERMISSION_SHOWCAMERA)) {",
+                    "           if (PermissionUtils.shouldShowRequestPermissionRationale(target.getActivity(), PERMISSION_SHOWCAMERA)) {",
```
Why this change is needed?

hotchemi commented:
@shiraji Your PR is almost awesome!

> To make it consist, local variable activity is now gone

You said it, but what do u mean?

shiraji commented:
Thanks

> You said it, but what do u mean?

`SupportFragmentProcessorUnit` used [local variable "activity"](https://github.com/hotchemi/PermissionsDispatcher/pull/65/files#diff-75ab769774b3ac028618505913ac9f8bL38), on the other hand, `NativeFragmentProcessorUnit` used [target.getActivity()](https://github.com/hotchemi/PermissionsDispatcher/pull/65/files#diff-de57b3433a39c68273e05dc9453aaf58L44). This difference was really hard to keep when I moved `addWithCheckBody()` to `BaseProcessorUnit`. Since, `ActivityProcessorUnit` [doesn't use local variable](https://github.com/hotchemi/PermissionsDispatcher/pull/65/files#diff-63fff1f7d89e03eddfd8caaad214dddaL34), to make it consist, I removed local variable "activity".

hotchemi commented:
ic.
I'll merge it after next version's release.
Thanks a lot!

shiraji commented:
Thanks for reviewing this. And I would like to thanks for giving me a chance to learn Kotlin!

hotchemi commented:
@shiraji It was the work of @aurae. I haven't done anything haha.
